### ISSUE TITLE
Replace or extend object-shorthand ignoreConstructors option with something more powerful

### ISSUE THREAD

**Issue by chris-morgan** [eslint/eslint#11686]
**What rule do you want to change?**

I want to replace the `object-shorthand` rule’s `ignoreConstructors` option with something more powerful in two ways:

- I want to be able to specify the rules for how it classifies whether or not something is a constructor (my specific use case being an old-style ES5 class creator which takes an `init` property which must be a constructor).
- I don’t want to *ignore* whether or not object shorthand is used on nominated constructors, but rather to produce an error if it *is* used, due to how shorthand methods lack [[Construct]].

**Does this change cause the rule to produce more or fewer warnings?**

No. Functionally it adds a new option. Should that option be used, it may produce more warnings because it allows you to complain about constructors that improperly use shorthand, or fewer warnings when you have a pattern of your own for constructor property names, because it allows you to classify them correctly.

**How will the change be implemented? (New option, new default behavior, etc.)?**

Define a new option for `object-shorthand`, named `constructors`. This property is an object with two optional keys:

- `pattern: string|string[]|RegExp|(string => boolean)` (default is the current `isConstructor` function, with its “starts with a capital letter” check): defines how to determine whether or not a method name should be considered a constructor. If it is a string or array of strings, the method name is tested for an exact match; if a RegExp, the method name is tested against it; if a function, it is called with the method name as its argument, and should return whether it is a constructor.
- `ignore: true` (default `false`): by default, any methods matching the constructor pattern will produce an error/warning if they use shorthand; setting this disables that warning.

Then deprecate `ignoreConstructors`, treating `ignoreConstructors: true` as equivalent to `constructors: { ignore: true }`.

There’s plenty of scope for bikeshedding. One variant I’ll propose is keeping `ignoreConstructors`, and making it imply `constructors: true` if `constructors` isn’t specified, and shifting what I called `constructors.pattern` up to `constructors`, with it also now permitting a boolean, `true` meaning to use the default classification function.

**Please provide some example code that this change will affect:**

```js
const GoodThing = Class({
    Extends: SomeThing,
    init: function () { },
    method () { },
});

const BadThing = Class({
    Extends: SomeThing,
    init () { },
    method () { },
});
```

(This example uses the Overture library that we use for making the FastMail web interface. `new BadThing()` will fail, since that `init` is supposed to be a constructor.)

**What does the rule currently do for this code?**

Complains that the `init` in `GoodThing` should use object shorthand.

**What will the rule do after it's changed?**

Combined with this config:

```js
'object-shorthand': ['error', {
    constructors: {
        pattern: 'init',
    },
}]
```

It will allow the `GoodThing` `init` function through, but complain that the `BadThing` `init` method should not be using shorthand.

**Are you willing to submit a pull request to implement this change?**

Yes.

**Comment by chris-morgan** [comment]
I do not believe that auto-closing is the appropriate action here: as I said, I’m willing to implement it myself. It’s purely a question of whether the proposed design is reasonable.

**Comment by kaicataldo** [comment]
Apologies for the lack of response. The team has been busy the last few months and this seems to have fallen through the cracks. I’m going to reopen this since we weren’t able to triage it the first time around.

Although the intended use is for constructors, it seems like it could be used for any arbitrary method name. I wonder if it would make sense to create a more generic API to match the functionality?

**Comment by chris-morgan** [comment]
Thanks.

I do not expect people will normally want to mix-and-match object shorthand; normally they will either want to use it everywhere, or not, everywhere, as a stylistic concern; constructors are the only exception that I can think of where the difference *matters*.

Given all this, I don’t know what more generic API you would consider.

**Comment by mysticatea** [comment]
Just a curiosity, what is the reason you don't use ES2015 classes even if you can use ES2015 (object shorthand)?

**Comment by chris-morgan** [comment]
@mysticatea Object shorthand is a syntax thing that can comparatively easily be adopted, but migrating from one style of classes to another is technically involved when there is not perfect overlap in supported features, which there isn’t. To make using `class` feasible and tolerable, we’d need the class field and decorator proposals, and even then, without decorators on object literals (which is currently slated to be future work) it’d be a pain. Switching from one style to the other would take a *lot* of work, and would be quite impossible to justify as it stands. (We’re a small team; at this time there are only two of us that work on the frontend.) For your interest, here’s a public example of some Overture code (written in an older style, as we haven’t updated it for how we currently write such code at FastMail): https://github.com/fastmail/overture/blob/c581a147b8452d8b6710e7f6257a59f4ff07cd9a/examples/Todo/Todo.js.
